### PR TITLE
Update AAD Token with the correct mHSM Resource URL

### PR DESCRIPTION
**Pull request by sid-habu**
For Service Principal based authentication to an mHSM, the scope was hardcoded to AKV.

### CODE REVIEW

File: cvm-securekey-release-app/AttestationUtil.cpp [hunk 1]
```diff
@@ -186,6 +186,34 @@ static inline std::string GetImdsTokenUrl(std::string url)
...
+    const std::string AKV_URL_SUFFIX = Constants::AKV_URL_SUFFIX;
+    const std::string MHSM_URL_SUFFIX = Constants::MHSM_URL_SUFFIX;
+    const std::string AKV_RESOURCE_URL = Constants::AKV_RESOURCE_URL;
+    const std::string MHSM_RESOURCE_URL = Constants::MHSM_RESOURCE_URL;
+
+    // Check if AKV suffix is present in KEKUrl
+    if (KEKUrl.find(AKV_URL_SUFFIX) != std::string::npos)
+    {
+        TRACE_OUT("AKV resource suffix found in KEKUrl");
+        return isIMDS ? AKV_RESOURCE_URL : AKV_RESOURCE_URL + "/.default";
```
Comment (sid-habu) [inline cvm-securekey-release-app/AttestationUtil.cpp:202]: AAD based auth requires `/.default` at the end of the URL in the scope

File: cvm-securekey-release-app/AttestationUtil.cpp [hunk 2]
```diff
@@ -198,22 +226,7 @@ std::string Util::GetIMDSToken(const std::string &KEKUrl)
...
-    {
-        resourceUrl = Constants::MHSM_RESOURCE_URL;
-        TRACE_OUT("MHSM resource suffix found in KEKUrl");
-    }
-    else
-    {
-        TRACE_ERROR_EXIT("Invalid resource suffix found in KEKUrl")
-    }
-
+    std::string resourceUrl = getResourceUrl(KEKUrl);
```
Comment (sid-habu) [inline cvm-securekey-release-app/AttestationUtil.cpp:229]: consolidated existing logic in a util method

File: cvm-securekey-release-app/AttestationUtil.cpp [hunk 3]
```diff
@@ -264,16 +277,17 @@ std::string Util::GetIMDSToken(const std::string &KEKUrl)
...
     TRACE_OUT("Entering Util::GetAADToken()");
 
     auto clientId = std::getenv("AKV_SKR_CLIENT_ID");
     auto clientSecret = std::getenv("AKV_SKR_CLIENT_SECRET");
     auto tenantId = std::getenv("AKV_SKR_TENANT_ID");
 
+    std::string resourceUrl = getResourceUrl(KEKUrl, false);
     std::string tokenUrl = "https://login.microsoftonline.com/" + std::string(tenantId) + "/oauth2/v2.0/token";
-    std::string postData = "client_id=" + std::string(clientId) + "&client_secret=" + std::string(clientSecret) + "&grant_type=client_credentials&scope=https://vault.azure.net/.default";
+    std::string postData = "client_id=" + std::string(clientId) + "&client_secret=" + std::string(clientSecret) + "&grant_type=client_credentials&scope= " + resourceUrl;
```
Comment (sid-habu) [inline cvm-securekey-release-app/AttestationUtil.cpp:290]: main change

**Comment by sid-habu** [comment]
@canfikret Can you please review this? I was able to test this in my fork using an mHSM and Service Principal based auth

**Comment by sid-habu** [comment]
@pankajosh @canfikret bumping this up

**Comment by sid-habu** [comment]
@canfikret Bumping this up. This has been running fine for us for 3 weeks. Let me know if you have questions

**Review by canfikret — APPROVED** [review]

**Pull request merged**